backend/storage/discussion-board.sql

```diff
@@ -0,0 +1,82 @@
+CREATE TYPE position_types AS ENUM ('Student', 'TA', 'Instructor');
+
+CREATE TABLE users (
+  user_id SERIAL PRIMARY KEY,
+  first_name varchar(50),
+  last_name varchar(50),
+  preferred_name varchar(50),
+  email varchar(320) UNIQUE NOT NULL
+);
+
+CREATE TABLE institutions (
+  id SERIAL PRIMARY KEY,
+  name varchar(100) NOT NULL,
+  location varchar(100)
+);
+
+CREATE TABLE semesters (
+  id SERIAL PRIMARY KEY,
+  semester_name varchar(100) NOT NULL,
+  institution_id int NOT NULL REFERENCES institutions (id) ON DELETE CASCADE
+);
+
+CREATE TABLE administrators (
+  user_id int REFERENCES users (user_id) ON DELETE CASCADE,
+  institution_id int REFERENCES institutions (id) ON DELETE CASCADE,
+  PRIMARY KEY (user_id, institution_id)
+);
+
+CREATE TABLE courses (
+  id SERIAL PRIMARY KEY,
+  course_name varchar(100),
+  course_code varchar(50) NOT NULL,
+  semester_id int REFERENCES semesters (id) ON DELETE CASCADE,
+  is_archived boolean
+);
+
+CREATE TABLE enrolment (
+  user_id int REFERENCES users (user_id) ON DELETE CASCADE,
+  course_id int REFERENCES courses (id) ON DELETE CASCADE,
+  position position_types NOT NULL,
+  PRIMARY KEY (user_id, course_id)
+);
+
+CREATE TABLE categories (
+  id SERIAL PRIMARY KEY,
+  category_name varchar(100) NOT NULL,
+  course int REFERENCES courses (id) ON DELETE CASCADE
+);
+
+CREATE TABLE posts (
+  id SERIAL PRIMARY KEY,
+  author int REFERENCES users (user_id) ON DELETE CASCADE,
+  parent_id int REFERENCES posts (id) ON DELETE CASCADE,
+  num_children int DEFAULT 0,
+  course int REFERENCES courses (id) ON DELETE CASCADE,
+  title varchar(200) NOT NULL,
+  body varchar,
+  img int,
+  upvotes int DEFAULT 0,
+  view_count int DEFAULT 0,
+  created_at timestamp with time zone DEFAULT (now()),
+  updated_at timestamp with time zone DEFAULT (now())
+);
```

Comment:
taking a look at the backlog, i think we could add a few fields for future features such as:
- is_locked (indicates whether a thread is no longer taking replies)
- is_anon (indicates whether we should hide the poster's identity)

some other considerations (that im not quite sure of but we should definitely discuss!):
- how will we handle a thread being "pinned"?
- will polls be considered a post/thread or require a different table?

Reply:
for 1), when a new course is added, we could always add a category with category_name "pinned" under that course? Threads with the "pinned" category can be displayed first.

Reply:
similar to is_anon, how are we determining anonymous names, I assume like Piazza we might want to generate it once per thread, if so, are we doing this deterministically (so we don't need to store any additional information), or are we doing it randomly, and we do need to store some attribute in the thread/comment?

Reply:
the latest commit has a separate table for polls, where the polls reference a post (or a top-level reply). The poll's votes are stored in yet another table, but I think this can allow for polls where people can click on multiple options